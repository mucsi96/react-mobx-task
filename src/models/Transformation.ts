import { observable } from 'mobx';

export default class Transformation {
  public readonly id: string;

  @observable
  public from: string;

  @observable
  public to: string;

  constructor(from: string, to: string) {
    this.id = Date.now().toString();
    this.from = from;
    this.to = to;
  }
}
