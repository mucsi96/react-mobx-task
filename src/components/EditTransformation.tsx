import * as React from 'react';
import { observable, action } from 'mobx';
import { inject, observer } from 'mobx-react';
import DictionaryStore from '../stores/DictionaryStore';
import ViewStore from '../stores/ViewStore';
import Dictionary from '../models/Dictionary';
import Transformation from '../models/Transformation';

interface EditTransformationProps {
  dictionaryId: string;
  transformationId?: string;
}

@inject('dictionaryStore', 'viewStore')
@observer
export default class EditTransformation extends React.Component<
  EditTransformationProps
> {
  get injected() {
    return (this.props as unknown) as {
      dictionaryStore: DictionaryStore;
      viewStore: ViewStore;
    };
  }

  @observable
  private from: string = '';

  @observable
  private to: string = '';

  private readonly dictionary: Dictionary;
  private readonly transformation?: Transformation;

  constructor(props: EditTransformationProps) {
    super(props);
    this.dictionary = this.injected.dictionaryStore.getDictionaryById(
      props.dictionaryId
    );
    if (props.transformationId) {
      this.transformation = this.dictionary.getTransformationById(
        props.transformationId
      );
      this.from = this.transformation.from;
      this.to = this.transformation.to;
    }
  }

  public render(): JSX.Element {
    return (
      <>
        <div className="ui basic clearing segment">
          <div className="ui left floated header">
            {this.transformation
              ? 'Edit transformation'
              : 'Create transformation'}
          </div>
          <div className="ui right floated header">
            {this.props.transformationId ? (
              <button
                className="ui red button"
                onClick={this.handleClickDelete}
              >
                Delete
              </button>
            ) : null}
          </div>
        </div>
        <div className="ui basic segment">
          <form className="ui form" onSubmit={this.handleSubmit}>
            <div className="field">
              <label>From</label>
              <input
                autoFocus={true}
                type="text"
                value={this.from}
                onChange={event => this.setFrom(event.target.value)}
                required={true}
              />
            </div>
            <div className="field">
              <label>To</label>
              <input
                type="text"
                value={this.to}
                onChange={event => this.setTo(event.target.value)}
                required={true}
              />
            </div>
            <button type="submit" className="ui button primary">
              {this.transformation ? 'Save' : 'Create'}
            </button>
            <button
              type="button"
              className="ui button"
              onClick={this.handleCancel}
            >
              Cancel
            </button>
          </form>
        </div>
      </>
    );
  }

  private handleSubmit = (event: React.FormEvent) => {
    event.preventDefault();
    if (!this.transformation) {
      this.dictionary.addTransformation(new Transformation(this.from, this.to));
    } else {
      this.transformation.from = this.from;
      this.transformation.to = this.to;
    }
    this.injected.viewStore.showDictionary(this.props.dictionaryId);
  };

  private handleCancel = () => {
    this.injected.viewStore.showDictionary(this.props.dictionaryId);
  };

  private handleClickDelete = () => {
    if (this.props.transformationId) {
      this.dictionary.removeTransformation(this.props.transformationId);
    }
    this.injected.viewStore.showDictionary(this.props.dictionaryId);
  };

  @action
  private setFrom(from: string) {
    this.from = from;
  }

  @action
  private setTo(to: string) {
    this.to = to;
  }
}
