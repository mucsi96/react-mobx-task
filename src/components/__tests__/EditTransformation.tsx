import * as React from 'react';
import * as RTL from 'react-testing-library';
import { Provider } from 'mobx-react';
import EditTransformation from '../EditTransformation';
import Transformation from '../../models/Transformation';

const testTransformation = { id: '1', from: 'b', to: 'c' };

const testDictionary = {
  getTransformationById: (id: string) =>
    id === '11' ? testTransformation : undefined,
  addTransformation: jest.fn(),
  removeTransformation: jest.fn()
};

const dictionaryStore = {
  getDictionaryById: (id: string) => (id === '7' ? testDictionary : undefined)
};

const viewStore = {
  showDictionary: jest.fn()
};

const render = ({ create = false } = {}) =>
  RTL.render(
    <Provider dictionaryStore={dictionaryStore} viewStore={viewStore}>
      <EditTransformation
        dictionaryId="7"
        transformationId={create ? undefined : '11'}
      />
    </Provider>
  );

describe('EditTransformation', () => {
  it('renders', () => {
    const { asFragment } = render();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders in creation case', () => {
    const { asFragment } = render({ create: true });
    expect(asFragment()).toMatchSnapshot();
  });

  it('modifies transformation on submit in case of editing', () => {
    const { getByLabelText, getByText } = render();
    RTL.fireEvent.change(getByLabelText('From'), {
      target: { value: 'new from' }
    });
    RTL.fireEvent.change(getByLabelText('To'), {
      target: { value: 'new to' }
    });
    RTL.fireEvent.submit(getByText('Save'));
    expect(testTransformation.from).toEqual('new from');
    expect(testTransformation.to).toEqual('new to');
  });

  it('creates new transformation on submit in case of creation', () => {
    testDictionary.addTransformation.mockClear();
    const { getByLabelText, getByText } = render({ create: true });
    RTL.fireEvent.change(getByLabelText('From'), {
      target: { value: 'new from' }
    });
    RTL.fireEvent.change(getByLabelText('To'), {
      target: { value: 'new to' }
    });
    RTL.fireEvent.submit(getByText('Create'));
    expect(testDictionary.addTransformation).toBeCalled();
  });

  it('navigates to dictionary view on "Cancel" click', () => {
    viewStore.showDictionary.mockClear();
    const { getByText } = render();
    RTL.fireEvent.click(getByText('Cancel'));
    expect(viewStore.showDictionary).toBeCalledWith('7');
  });

  it('removes transformation on "Delete" click', () => {
    testDictionary.removeTransformation.mockClear();
    const { getByText } = render();
    RTL.fireEvent.click(getByText('Delete'));
    expect(testDictionary.removeTransformation).toBeCalledWith('11');
  });

  it('navigates to dictionary view on "Delete" click', () => {
    viewStore.showDictionary.mockClear();
    const { getByText } = render();
    RTL.fireEvent.click(getByText('Delete'));
    expect(viewStore.showDictionary).toBeCalledWith('7');
  });
});
