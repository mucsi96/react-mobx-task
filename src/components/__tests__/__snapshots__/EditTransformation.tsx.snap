// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditTransformation renders 1`] = `
<DocumentFragment>
  <div
    class="ui basic clearing segment"
  >
    <div
      class="ui left floated header"
    >
      Edit transformation
    </div>
    <div
      class="ui right floated header"
    >
      <button
        class="ui red button"
      >
        Delete
      </button>
    </div>
  </div>
  <div
    class="ui basic segment"
  >
    <form
      class="ui form"
    >
      <div
        class="field"
      >
        <label
          for="from"
        >
          From
        </label>
        <input
          id="from"
          required=""
          type="text"
          value="b"
        />
      </div>
      <div
        class="field"
      >
        <label
          for="to"
        >
          To
        </label>
        <input
          id="to"
          required=""
          type="text"
          value="c"
        />
      </div>
      <button
        class="ui button primary"
        type="submit"
      >
        Save
      </button>
      <button
        class="ui button"
        type="button"
      >
        Cancel
      </button>
    </form>
  </div>
</DocumentFragment>
`;

exports[`EditTransformation renders in creation case 1`] = `
<DocumentFragment>
  <div
    class="ui basic clearing segment"
  >
    <div
      class="ui left floated header"
    >
      Create transformation
    </div>
    <div
      class="ui right floated header"
    />
  </div>
  <div
    class="ui basic segment"
  >
    <form
      class="ui form"
    >
      <div
        class="field"
      >
        <label
          for="from"
        >
          From
        </label>
        <input
          id="from"
          required=""
          type="text"
          value=""
        />
      </div>
      <div
        class="field"
      >
        <label
          for="to"
        >
          To
        </label>
        <input
          id="to"
          required=""
          type="text"
          value=""
        />
      </div>
      <button
        class="ui button primary"
        type="submit"
      >
        Create
      </button>
      <button
        class="ui button"
        type="button"
      >
        Cancel
      </button>
    </form>
  </div>
</DocumentFragment>
`;
